From 8c24320584eb2bee632a217620cfe90ea3d5c015 Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sat, 27 Sep 2014 21:03:55 +0800
Subject: [PATCH 0124/5242] MLK-9638-3 usb: chipidea: udc: test: handle
 otg_hnp_req set feature request

commit  8d032a0034d123da618c1fe48dedbd718f3f334b from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds otg_hnp_reqd set feature request from A host for chipidea udc
driver.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 450d4a1eb40ec4284661c1130bdcc4f4d2e6c527)
(cherry picked from commit 03f3f886029fa8f25dde34fcfeb28589ead0064f)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/udc.c |   19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 5e80b80..23592c3 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -749,8 +749,10 @@ static void isr_reset_handler(struct ci_hdrc *ci)
 {
 	int retval;
 
-	if (ci_otg_is_fsm_mode(ci))
+	if (ci_otg_is_fsm_mode(ci)) {
 		ci->fsm.otg_srp_reqd = 0;
+		ci->fsm.otg_hnp_reqd = 0;
+	}
 
 	spin_unlock(&ci->lock);
 	if (ci->gadget.speed != USB_SPEED_UNKNOWN)
@@ -1027,6 +1029,18 @@ static int otg_srp_reqd(struct ci_hdrc *ci)
 	}
 }
 
+static int otg_hnp_reqd(struct ci_hdrc *ci)
+{
+	if (ci_otg_is_fsm_mode(ci)) {
+		ci->fsm.otg_hnp_reqd = 1;
+		ci->fsm.b_bus_req = 1;
+		ci->gadget.host_request_flag = 1;
+		return isr_setup_status_phase(ci);
+	} else {
+		return -ENOTSUPP;
+	}
+}
+
 /**
  * isr_setup_packet_handler: setup packet handler
  * @ci: UDC descriptor
@@ -1153,6 +1167,9 @@ static void isr_setup_packet_handler(struct ci_hdrc *ci)
 				case TEST_OTG_SRP_REQD:
 					err = otg_srp_reqd(ci);
 					break;
+				case TEST_OTG_HNP_REQD:
+					err = otg_hnp_reqd(ci);
+					break;
 				default:
 					break;
 				}
-- 
1.7.9.5

